### PR TITLE
Rework code review section

### LINKED ISSUE #217: Make the code review section more engaging
Currently the code review section ends with a long bullet point list of best practices. I propose restructuring this section:

* Go through technical steps to raise a review
* Get user to raise a review based on an example branch that will be included in the repo template
* Go through technical steps of adding a comment, get user to add a comment
* Cover some of the things that a review should look for
* Get user to go through the code looking for those things
* Repeat last two points if there are more than 6 things to look for
* Suggest a different person then attempts to address their review comments (or address their own if doing on own)
* Talk about empathy (and any other review considerations)

Steps to do this:
 
 - [x] Identify the kind of things to look for in a review
 - [x]  Propose new structure for the section
 - [x] Develop a piece of code that exhibits the problems
 - [x] Write the exercises and example solutions
 - [ ] Add new conent
 - [ ] Modify the content for creating a new pull request (this is quite different)
 - [ ] Modify existing content
 - [ ] Read through and make final tweaks
(I intend to address this, raising issue for visibility and in case anyone disagrees).

### PR DESCRIPTION
**Pull request by thomaskileyukaea**
Changes to the code review section, fixes #217 

The exercises are going to have the reader be reviewing a pull request identical to this one: https://github.com/thomaskileyukaea/python-intermediate-inflammation/pull/3

One major outstanding issue is how to get the example branch onto the users repo (templates don't copy branches in a way that would work). 

It may be easiest to see the changes commit by commit, as I did the structural changes first, before adding new things. 

- [x]  Rewrite ` Code Review and Pull Requests In Action` for the new exercise
- [x]  Update branch diagram for new branch situation (or remove)
- [x] Update screen shot for the new pull request
- [x] Update `Raising a pull request` for the new exercise
- [x] Screenshot todos
- [x] Fix numbering
- [x] Tidy up or remove `Best Practice for Code Review` 
- [x] Line by line edit of new content
- [x] SemBr pass
- [x] Final read through

### CODE REVIEW

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:93]: I didn't think this exercise added much and I wanted to add other more exercises - but I'm happy to discuss.

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:154]: I don't see how this section contributes to the goals, so removing this gives more time to focus on the kind of review system we're recommending - tool assisted review.

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:272]: I restructured the examples so they don't require adding collaborators. I would be tempted to move this into the extras though, as it might be useful in some instances to know how to do this.

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:310]: Now the exercises are reviewing a fixed changeset this is no longer required.

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:345]: As above - reader is no longer required to write some tests for someone else to review

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:381]: ❓ I'm not sure this is actually that useful so I'm tempted to just cut it.

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:543]: Here I am assuming the reader is part of a team. However, I remember in discussions we talked how lots of users of this course are working independently. How do those scientists who are working independently use the code review section? Do they find someone else in their dept to review their code, reviewing it themselves, or just skipping this section?

**Review by thomaskileyukaea — COMMENTED** [review]
Explanation of why I removed certain sections / exercises as well as some ❓ questions for the maintainers that we should think about before merging this.

File: _episodes/41-code-review.md
Comment (bielsnohr) [reply _episodes/41-code-review.md:381]: I sort of like this because it helps focus what code review should be about. That being said, it
isn't as important as the previous tips about what to look for, so perhaps we assess this based on
how long the section takes to deliver?

File: _episodes/41-code-review.md
Comment (bielsnohr) [inline _episodes/41-code-review.md:416]: Will need to get rid of "Step 6" and make sure the heading level is in line with the above.

File: _episodes/41-code-review.md
Comment (bielsnohr) [inline _episodes/41-code-review.md:501]: The paragraphs need some semantic breaks at periods.

File: _episodes/41-code-review.md
Comment (bielsnohr) [inline _episodes/41-code-review.md:510]: Sembr

File: _episodes/41-code-review.md
Comment (bielsnohr) [reply _episodes/41-code-review.md:543]: This is a good question. As someone who has lone coded in various scientific settings, I still use
merge/pull requests as a gate check on what I have just done. Of course, I haven't done that in
every instance, but it is a nice idea to strive for.

In terms of actual code review, this is tough. I was part of a community effort to try and define
what code review in research might look like: https://dev-review.readthedocs.io/en/latest/index.html

There, we did recommend trying to find someone to review your code. That person could be in your
research group (but just not working directly on what you are). We could link to that page here do
deal with the case of solo researchers working on code.

**Review by bielsnohr — COMMENTED** [review]
Overall, I really like where this is heading. The exercise based on a pre-determined changeset should help better achieve the objectives of this section. In the previous version, not having control over the incoming changes meant that learners might not see many of the points being raised in this section. So, this new approach gets my full endorsement.

It will be great to see how the section flows once the wording around this new exercise is added.

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:154]: I will make this adjustment as I think you're right, it is nice to have some other terms to look up, but with less detail.

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md]: Yes - this is now done.

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:186]: Make these instructions go to the pull requests page and specify the branches rather than relying on the time-sensitive button to appear.

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:193]: Maybe moving this content to after explaining what to look for would encourage the correct flow

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:262]: Explicitly show before hand how to add comments, and make explicit they should be added to the review but not submitted yet

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:237]: Could also talk about consistency of code here

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:185]: Here and elsewhere, github now refers to the head branch as the compare branch.

**Review by thomaskileyukaea — COMMENTED** [review]
Feedback from running through this with @bielsnohr Also noting we got to and mostly did responding to comments and it looks like it is 25min reading, 30min of exercises up to there. If the numbers at the top are accurate, it would seem this would take a little longer

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:314]: First use of PR should be explained

**Comment by thomaskileyukaea** [comment]
Finished off, looks like another 5min of reading, plus another 10 min of exercise.

File: _episodes/41-code-review.md
Comment (anenadic) [inline _episodes/41-code-review.md:256]: Feels like we need "what steps are of the code review?" to highlight the process earlier in the episode. So, e.g. reviewing the code and what to look for, commenting on code, buffering all comments and finishing/sending off our review, responding to review comments. Or something like that.

File: _episodes/41-code-review.md
Comment (anenadic) [reply _episodes/41-code-review.md:245]: Also, "Effective review comments" could go under "adding a comment" either as the main text or as a callout.

File: _episodes/41-code-review.md
Comment (anenadic) [inline _episodes/41-code-review.md:423]: This section could be formatted as a bullet point list.

**Review by anenadic — COMMENTED** [review]
Thank you for the PR @thomaskileyukaea - apologies if took a while to review. We had a maintainers' meeting last week where we reviewed your changes. There is a couple of comments that can be addressed now (see my review of your PR) and then we'd like to have another meeting with you to discuss bigger changes which may be more difficult to explain in the PR's comments (and also because it affects some earlier episodes). If you are not available to meet and to work on this - we'd be happy to accept the PR to the dev branch and then apply our changes on top of yours, if that is OK with you.

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:68]: Incorporated the question and some of the solution into the final exercise to reinforce the key messages of the lesson.

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:176]: Unfortunately we believe the template approach is a bit of a non-starter for this due to the branch history being lost, meaning it isn't possible to create a PR from one branch to the other. Matt will discuss this with you in more detail at the next meeting to work out the best way forward. (I sadly won't be able to make it).

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:195]: I'm not sure I fully understand this comment, sorry. 

To add to the confusion, I recently changed how the collaborative code development models are structured on the main branch, this isn't merged into this branch(!). 

It is currently structured like:

> * Is the code readable
> * Is the code a minimal change
> ...
> 
> ### Is the code readable
> 
> More details about what this means
>
> ### Is the code a minimal change
>
> More details about what this means. 

Which, to my eyes, is the same as the collaborative code development models on the main branch. 

We _could_ structure like this:

> * Is the code readable - more details about what this means
> * Is the code a minimal change - more details about what this means
> ...

But to my eyes that is less clear as the bullet points become so long as to not really feel like bullet points. 

Alternatives I can see:

1. Make the bullet point list link to the sections so it is clear to the reader there is more detail to come
2. Make the bullet point list include a longer summary of each of the points, but still have the full sections after
3. Make the headings the bullet point list refers to smaller (e.g. just bold) so it looks more connected. 

Do let me know which of these you think would be best (or some combination of them). 

I'm not sure about the title change. I like that the use of the word "aim" emphasises that we're trying to get the code to a specific quality, but I worry it makes it ambiguous whether we're thinking about this from the reviewer or the authors perspective. How about:

* What should a code review aim to improve
* What are the qualities a code review should be aiming to improve

We do introduce the idea of design patterns in section 3, but I think modularity is a better thing to focus on anyway!

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:285]: I actually carefully setup this exercise to _not_ require collaborators (as you end up only making changes on your own repo, and making comments [which doesn't require collaborator permissions] on the other peoples). That said, it maybe is a bit complicated to see that - will add a diagram which will hopefully make it clearer.

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:195]: I have ended up doing at least (3) (though this will make 1 harder to do) as I think it helps with clarity.

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:245]: I've added

> Once a review has been raised it is over to the reviewer to review the code and submit a review.

At the start of the section on reviewing a pull request - is that what you meant?

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:256]: I've added a diagram in the code review in action section which will hopefully address this comment - let me know if you think we need more.

File: _episodes/41-code-review.md
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:350]: I opted for putting the different review responses as a bullet point list, which I think makes sense, but maybe makes making the steps a list a bit much - let me know what you think

**Comment by thomaskileyukaea** [comment]
Thanks @anenadic for the review - it was really helpful. I've tried to address your comments - or else have replied with further clarifications. Do let me know if you have any further feedback.

I've merged the latest gh-pages into this so that it has the same collaborative code models.

**Comment by thomaskileyukaea** [comment]
I also looked into using Mermaid diagrams directly - at first it seemed tricky, but ended up being quite straight forward - here's a PR on my fork showing how it could be done: https://github.com/thomaskileyukaea/python-intermediate-development/pull/1

If we did go down this route, we could convert a few of the git diagrams into mermaid diagrams too, which would make maintenance easier.

**Comment by bielsnohr** [comment]
Comments from maintainer sprint:

- Concern about confusion when initially forking and having multiple branches. Solution: add some text around initial forking instructions about ignoring the additional branches
- The reasons for forking (TODO fill out) (i.e. need for retaining history in order to raise pull request)
- Options to retain templating:
  1. GitHub action that creates the required branch and changes upon creation of the copy from the template
  2. Have a script in the template repo that learners would run themselves (downside: system dependencies and learner error)
- obvious downsides are developer time to get these up and running, and then maintenance costs

**Comment by bielsnohr** [comment]
> I also looked into using Mermaid diagrams directly - at first it seemed tricky, but ended up being quite straight forward - here's a PR on my fork showing how it could be done: [thomaskileyukaea#1](https://github.com/thomaskileyukaea/python-intermediate-development/pull/1)
> 
> If we did go down this route, we could convert a few of the git diagrams into mermaid diagrams too, which would make maintenance easier.

Aha! Great that you were able to get this working. Of course, I should have remembered the solution of just injecting the required Javascript somewhere in the HTML. One gets quite used to everything being done with packages/gems/etc...

I would vote that we do go down the route of adding Mermaid. The diagram for this lesson looks great, see the bottom of this comment. GitHub already supports Mermaid, so the below is actually a rendering of the source code. Also, as @thomaskileyukaea mentions, we can convert other diagrams to Mermaid and it will make maintenance of the website easier (no more fiddling with svg, regenerating a png/jpeg, making sure they are in the right folder, etc).

Could I request that we get a maintainer vote on this? @anenadic I'll have a look through the governance document we are working on, but email list probably best?

```mermaid
%% Must disable useMaxWidth to have the diagram centre
%%{init: { 'sequence': {'useMaxWidth':false} } }%%
sequenceDiagram
    participant A as Author
    participant R as Reviewer
    A->>A: Write some code
    A->>R: Raise a pull request
    R->>R: Add comments to a review
    R->>A: Submit a review
    loop Until approved
        A->>R: Address or respond to review comments
        R->>A: Clarify or resolve comments
    end
    R->>A: Approve pull request
    A->>A: Merge pull request
```

**Comment by bielsnohr** [comment]
> we'd like to have another meeting with you to discuss bigger changes which may be more difficult to explain in the PR's comments (and also because it affects some earlier episodes). If you are not available to meet and to work on this - we'd be happy to accept the PR to the dev branch and then apply our changes on top of yours, if that is OK with you.

@anenadic I think because of various AL schedules, it has been difficult to arrange a synchronous call about the bigger changes you mention above. By "bigger changes" are you referring to the fact that steps in some of the initial episodes for obtaining the project would need to be modified?

If that is the extent of the changes, then I think it is quite manageable. There is the issue of regenerating some images of the GitHub interface, but I can assist with that.

As you suggest, it is probably best that we accept this MR in its current state into the `dev` branch and then continue working on the other changes earlier in the course ourselves (i.e. maintainers). I am keen to have Thomas get going on his proposed modifications to the Architecture and Design sections.

**Review by bielsnohr — APPROVED** [review]
As mentioned in comments above, I think we are fine to merge this into the dev branch and then take some time to make the edits to earlier parts of the lesson to advise on using forking.

**Comment by thomaskileyukaea** [comment]
✔️ Addressed issue where the schedule appeared at the bottom after the mermaid change.

**Comment by anenadic** [comment]
> Summary of outcome from Maintainer Meeting 2023-10-25:
> 
> * @anenadic and @steve-crouch have communicated that there is some content from the original version of the episode that they don't want to see removed, as has been done here
>   
>   * with @thomaskileyukaea having no more immediate time to work on this, the maintainers will take on the task of making any changes to get this updated episode to a state where it is ready to merge
> * it was decided that it probably isn't a good idea to merge this into `dev` since there are still things that need to be reviewed and changes to be made; therefore, this PR will stay open for the immediate future while this work is done
> * @bielsnohr will take on the task of modifying the earlier part of the lesson where a fork is now required, as stated in [Change initial retrieval of course project to be a fork rather than template copy #249](https://github.com/carpentries-incubator/python-intermediate-development/issues/249) .
>   
>   * this work will be pushed to a separate branch that stacks on top of this one

We can either re-create a dev branch for this PR or have a new branch of the same name `rework-review-section`. Will try to find some time to work on this with @steve-crouch - hopefully tomorrow. So we can have this merged as soon as possible. Apologies for not responding sooner - things have gone a bit hectic.

**Comment by bielsnohr** [comment]
> We can either re-create a dev branch for this PR or have a new branch of the same name `rework-review-section`. Will try to find some time to work on this with @steve-crouch - hopefully tomorrow. So we can have this merged as soon as possible. Apologies for not responding sooner - things have gone a bit hectic.

I think creating a new branch within our repo to hold this work is the best next step. I'll do that shortly. 

It would be good if we could collect the concrete items that need to change in this issue: #217

**Pull request merged**
